pgxn/neon/console_connector.c

```diff
@@ -0,0 +1,632 @@
+        DbEntry *entry;
+        hash_seq_init(&status, RootTable.db_table);
+        while((entry = hash_seq_search(&status)) != NULL)
+        {
+            pushJsonbValue(&state, WJB_BEGIN_OBJECT, NULL);
+            PushKeyValue(&state, &estimated_len, "op", entry->type == Op_Set ? "set" : "del");
+            PushKeyValue(&state, &estimated_len, "name", entry->name);
+            PushKeyValue(&state, &estimated_len, "type", "db");
+            if(entry->owner[0] != '\0')
+            {
+                PushKeyValue(&state, &estimated_len, "owner", entry->owner);
+            }
+            if(entry->old_name[0] != '\0')
+            {
+                PushKeyValue(&state, &estimated_len, "old_name", entry->old_name);
+            }
+            pushJsonbValue(&state, WJB_END_OBJECT, NULL);
+            estimated_len += 3; // Add 3 for two curly braces and a comma
+        }
+    }
+
+    if(RootTable.role_table)
+    {
+        HASH_SEQ_STATUS status;
+        RoleEntry *entry;
+        hash_seq_init(&status, RootTable.role_table);
+        while((entry = hash_seq_search(&status)) != NULL)
+        {
+            pushJsonbValue(&state, WJB_BEGIN_OBJECT, NULL);
+            PushKeyValue(&state, &estimated_len, "op", entry->type == Op_Set ? "set" : "del");
+            PushKeyValue(&state, &estimated_len, "name", entry->name);
+            PushKeyValue(&state, &estimated_len, "type", "role");
+            if(entry->password)
+            {
+                PushKeyValue(&state, &estimated_len, "password", (char *)entry->password);
+            }
+            if(entry->old_name[0] != '\0')
+            {
+                PushKeyValue(&state, &estimated_len, "old_name", entry->old_name);
+            }
+            pushJsonbValue(&state, WJB_END_OBJECT, NULL);
+            estimated_len += 3;
+        }
+    }
+    JsonbValue *result = pushJsonbValue(&state, WJB_END_ARRAY, NULL);
+    Jsonb *jsonb = JsonbValueToJsonb(result);
+    return JsonbToCString(NULL, &jsonb->root, estimated_len);
+}
+
+static void SendDeltasToConsole()
+{
+    if(!RootTable.db_table && !RootTable.role_table)
+        return;
+    if(!ConsoleURL)
+    {
+        elog(LOG, "ConsoleURL not set, skipping forwarding");
+        return;
+    }
+    CURL *curl = curl_easy_init();
+    if(!curl)
+    {
+        elog(ERROR, "Failed to initialize curl object");
+        return;
+    }
+
+    char *message = ConstructDeltaMessage();
+
+    struct curl_slist *header = NULL;
+    header = curl_slist_append(header, "Content-Type: application/json");
+    curl_easy_setopt(curl, CURLOPT_CUSTOMREQUEST, "PATCH");
+    curl_easy_setopt(curl, CURLOPT_HTTPHEADER, header);
+    curl_easy_setopt(curl, CURLOPT_POSTFIELDS, message);
+    curl_easy_setopt(curl, CURLOPT_URL, ConsoleURL);
+
+    const int num_retries = 5;
+    int curl_status;
+    for(int i = 0; i < num_retries; i++)
+    {
+        if((curl_status = curl_easy_perform(curl)) == 0)
+            break;
+        pg_usleep(1000);
```

Comment:
This is 1 millisecond. Seems short, did you intend 1 second?

Reply:
Ah yes I did, good catch